test/test_literal.py

```diff
@@ -0,0 +1,174 @@
+
+
+def test_literal_argument_int():
+    @dataclass
+    class SomeClassInt:
+        a: Literal[123, 456]
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClassInt, dest="some_class")
+
+    args = parser.parse_args(shlex.split("--a 123"))
+    assert args == argparse.Namespace(some_class=SomeClassInt(a=123))
+
+    with exits_and_writes_to_stderr("invalid choice"):
+        parser.parse_args(shlex.split("--a 789"))
+
+    with raises(SystemExit):
+        parser.parse_args(shlex.split("--a test"))
+
+    with raises_missing_required_arg():
+        parser.parse_args("")
+
+
+def test_literal_optional():
+    @dataclass
+    class SomeClass:
+        a: Optional[Literal[123, 456]] = 123
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest="some_class")
+
+    args = parser.parse_args(shlex.split(""))
+    assert args == argparse.Namespace(some_class=SomeClass(a=123))
+
+    args = parser.parse_args(shlex.split("--a"))
+    assert args == argparse.Namespace(some_class=SomeClass(a=None))
+
+    args = parser.parse_args(shlex.split("--a 456"))
+    assert args == argparse.Namespace(some_class=SomeClass(a=456))
+
+    # args = parser.parse_args(shlex.split('--a None'))
+    # assert args == argparse.Namespace(some_class=SomeClass(a=None))
+    # args = parser.parse_args(shlex.split('--a none'))
+    # assert args == argparse.Namespace(some_class=SomeClass(a=None))
+
+
+def test_literal_optional_without_default():
+    @dataclass
+    class SomeClass:
+        a: Optional[Literal[123, 456]]
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest="some_class")
+
+    args = parser.parse_args(shlex.split(""))
+    assert args == argparse.Namespace(some_class=SomeClass(a=None))
+
+    args = parser.parse_args(shlex.split("--a"))
+    assert args == argparse.Namespace(some_class=SomeClass(a=None))
+
+    args = parser.parse_args(shlex.split("--a 123"))
+    assert args == argparse.Namespace(some_class=SomeClass(a=123))
+
+    # args = parser.parse_args(shlex.split('--a None'))
+    # assert args == argparse.Namespace(some_class=SomeClass(a=None))
+    # args = parser.parse_args(shlex.split('--a none'))
+    # assert args == argparse.Namespace(some_class=SomeClass(a=None))
+
+
+@parametrize(
+    "list_formatting_function",
+    [
+        format_list_using_spaces,
+        # format_list_using_brackets,
+        xfail_param(
+            format_list_using_brackets,
+            reason="TODO: decide which syntax we want to allow for single lists.",
+        ),
+        xfail_param(
+            format_list_using_double_quotes,
+            reason="TODO: decide which syntax we want to allow for single lists.",
+        ),
+        xfail_param(
+            format_list_using_single_quotes,
+            reason="TODO: decide which syntax we want to allow for single lists.",
+        ),
+    ],
+)
+def test_literal_argument_list(list_formatting_function):
+    @dataclass
+    class SomeClassInt:
```

Comment:
Please use the `TestSetup` class as a base class for your inner classes. This will make all the checks everywhere a lot easier to read:
Before:
```python
parser = ArgumentParser()
parser.add_arguments(SomeClassInt, dest="some_class")
args = parser.parse_args(shlex.split(f"--a 123"))
assert args == argparse.Namespace(some_class=SomeClassInt(a=[123]))
```
After:
```python
assert SomeClassInt.setup("--a 123") == SomeClassInt(a=[123])
```